pilot/pkg/bootstrap/configcontroller.go

```diff
@@ -268,6 +276,41 @@ func mcpSecurityOptions(ctx context.Context, configSource *meshconfig.ConfigSour
 	return securityOption, nil
 }
 
+// initInprocessAnalysisController spins up an instance of Galley which serves no purpose other than
+// running Analyzers for status updates.  The Status Updater will eventually need to allow input from istiod
+// to support config distribution status as well.
+func (s *Server) initInprocessAnalysisController(args *PilotArgs) error {
+
+	processingArgs := settings.DefaultArgs()
+	processingArgs.KubeConfig = args.Config.KubeConfig
+	processingArgs.EnableValidationController = false
+	processingArgs.EnableValidationServer = false
+	processingArgs.MonitoringPort = 0
+	processingArgs.Liveness.UpdateInterval = 0
+	processingArgs.Readiness.UpdateInterval = 0
+	processingArgs.Insecure = true // TODO - use sidecar?
+	processingArgs.EnableServer = false
+	processingArgs.MeshConfigFile = args.Mesh.ConfigFile
+	processingArgs.EnableConfigAnalysis = true
+
+	processing := components.NewProcessing(processingArgs)
+
+	s.leaderElection.AddRunFunction(func(stop <-chan struct{}) {
+		if err := processing.Start(); err != nil {
+			log.Errorf("error starting galley analysis server: %v", err)
+			// TODO: if we can't start analysis, we should un-elect ourselves, but pilot's API doesn't provide that
+			// should we add an unelect interface to pilot?
```

Comment:
This concerns me.  And the stop ctx for leader election is the istiod server stop context, so the only way to release the leader lock is to die altogether...

Reply:
The context is re-created after election is lost (see https://github.com/istio/istio/blob/5ea7387905082ab0374d4184b06deced86d72ba4/pilot/pkg/leaderelection/leaderelection.go#L55). I think the only change required is to pass the context (or the cancel function) to the leader election run functions so that they have the option to give up election.

Reply:
I think with everything else we just exit if we fail on startup, not sure we need to do anything fancy here?

Reply:
Should I just write to stop?

Reply:
@ayj unfortunately, the cancel function does not cause the leader lock to be released.  See https://github.com/istio/istio/blob/5ea7387905082ab0374d4184b06deced86d72ba4/pilot/pkg/leaderelection/leaderelection.go#L103

If I am reading this correctly, this also means that pilot never gracefully releases the leader lock.  When the leader shuts down (even gracefully), it simply stops renewing the lease, and 30 seconds may elapse before the lease expires and another leader takes its place.  Am I understanding this correctly?

Reply:
Ah, then I'd just exit with the appropriate error message and revisit later when ReleaseOnCancel is fixed.